File: medium/0994_rotting_oranges/solution.cpp
#include <set>
#include <vector>

class Solution {
public:
  int orangesRotting(std::vector<std::vector<int>> &grid) {
    const int rows = grid.size();
    const int cols = grid[0].size();

    std::vector<std::pair<int, int>> rotten;
    int freshCount = 0;

    for (int row = 0; row < rows; ++row) {
      for (int col = 0; col < cols; ++col) {
        if (grid[row][col] == 2)
          rotten.push_back({row, col});
        else if (grid[row][col] == 1)
          ++freshCount;
      }
    }

    int minutes = 0;

    while (!rotten.empty() && freshCount) {
      std::set<std::pair<int, int>> toRot;

      for (const auto &[row, col] : rotten) {
        if (row > 0 && grid[row - 1][col] == 1)
          toRot.insert({row - 1, col});
        if (col + 1 < cols && grid[row][col + 1] == 1)
          toRot.insert({row, col + 1});
        if (row + 1 < rows && grid[row + 1][col] == 1)
          toRot.insert({row + 1, col});
        if (col > 0 && grid[row][col - 1] == 1)
          toRot.insert({row, col - 1});
      }

      rotten = std::vector<std::pair<int, int>>(toRot.begin(), toRot.end());
      for (const auto &[row, col] : toRot) {
        grid[row][col] = 2;
        --freshCount;
      }

      ++minutes;
    }

    return freshCount == 0 ? minutes : -1;
  }
};

File: medium/0994_rotting_oranges/solution.py
from typing import List


class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        minutes = 0

        rotted = []
        to_rot = [
            (row, col)
            for row in range(len(grid))
            for col in range(len(grid[0]))
            if grid[row][col] == 2
        ]

        while to_rot:
            rotted.extend(to_rot)
            to_rot = []

            while rotted:
                row, col = rotted.pop()

                if row > 0 and grid[row - 1][col] == 1:
                    to_rot.append((row - 1, col))
                if col + 1 < len(grid[0]) and grid[row][col + 1] == 1:
                    to_rot.append((row, col + 1))
                if row + 1 < len(grid) and grid[row + 1][col] == 1:
                    to_rot.append((row + 1, col))
                if col > 0 and grid[row][col - 1] == 1:
                    to_rot.append((row, col - 1))

            if not to_rot:
                break

            to_rot = list(set(to_rot))
            for row, col in to_rot:
                grid[row][col] = 2

            minutes += 1

        for row in grid:
            if 1 in row:
                return -1

        return minutes
